### ISSUE TITLE
compiling templates with partials does not detect cycles

### ISSUE THREAD

**Issue by asariley** [JustusAdam/mustache#22]
While the spec does say "Don't make cycles in your partials" it would be better to error out somehow than to loop infinitely.

**Comment by JustusAdam** [comment]
That's difficult though because partials can be conditional and the conditional could be scoped and therefore change value in each partial call.

Mustache may be too strong as a language. Properly detecting such cycle is probably close to solving the halting problem.

**Comment by asariley** [comment]
Well currently, because the compilation process expands all partials, so this type of conditional nuance is not supported anyways. The spec says that partials should be expanded at run time which unfortunately seems to imply `IO` is needed at substitution time. 

I think there may be a slight tweak to the compilation process that will allow it to support these kind of runtime cycles. But I need to explore the idea a bit more. Failing that, I think the recursion depth idea is a good way to mitigate the infinite loop issue.

**Comment by JustusAdam** [comment]
Yes partials are expanded at runtime, however the partials are (by default) read at compilation time.
I know this is slightly against the proposed semantics and now that I say that it might be interesting to offer a different expansion strategy where it compiles templates at expansion time, but that is a topic for another issue.

I understand what you mean with the infinite loop now. The compilation step does some state propagation to avoid recompiling templates, perhaps this could be tweaked to insert the template *before* it has actually been compiled.

**Comment by JustusAdam** [comment]
When I designed this I went for ease of use, and that's why the partials are bundled with the Template itself. The recursive problem could be solved by splitting apart partials cache and templates and using a flat cache rather than a nested one.
